### PR TITLE
Fixes #378 - Plugin administration UI.

### LINKED ISSUE #378: Plugin administration UI
What it should do:

  * display all plugins names.
  * display plugins description: which need to be define/extract somehow.
  * enable/disable plugins.
  * setup plugin configuration.

### PR DESCRIPTION
**Pull request by ArthurHoaro**
## Features

From the administration panel, users can:

 * See all plugin name and description available.
 * Enable/disable plugins.
 * Change loading order of their plugins (affects for example shaares icon order).
 * Affect and change plugins parameters.

## Plugin developer side

Every plugin needs a `<plugin_name>.meta` file which is in fact an `.ini` file (`KEY="VALUE"`) to be listed in plugin administration.

Each file contain two keys:

  * `description`: plugin description
  * `parameters`: user parameter names, separated by a `;`.

> Note: In PHP, `parse_ini_file()` seems to want strings to be between by quotes `"` in the ini file.

## Theme maintainers

There is a new page: `pluginsadmin.html`. It needs to be added to your theme. I'll update documentation later.

Aside from classic RainTPL loops, plugins order is handle by JavaScript. You can use `plugin_admin.js`, only if:

  * you're using a table.
  * you call `orderUp()` and `orderUp()` onclick on arrows.
  * you add `data-line` and `data-order` to your rows.

Otherwise, theme maintainers can use their own JS as long as this field is send by the form:

    <input type="hidden" name="order_{$key}" value="{$counter}">

## Other

  * In `tools.html`, moved "Change password" after Shaarli and plugins configuration.
  * Unit tests.
  * 2 new page in `Router`: `pluginadmin` and `save_pluginadmin`.

## Preview

screenshot

## TODO

- [x] Documentation!

### CODE REVIEW

**Comment by virtualtam** [comment]
Neat \o/

Gonna start to test it during the upcoming weekend!

**Comment by ArthurHoaro** [comment]
Rebased.

File: tpl/pluginsadmin.html [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+<noscript>
+<div>
+    <ul class="errors">
+        <li>You need to enable Javascript to change plugin loading order.</li>
+    </ul>
+</div>
+<div class="clear"></div>
+</noscript>
+
+<div id="pluginsadmin">
```
Comment (virtualtam) [inline tpl/pluginsadmin.html]: though 4 spaces are a common indentation value for most languages, HTML devs tend to stick to using 2 spaces to keep the overall indentation readable:
- [w3schools](http://www.w3schools.com/html/html5_syntax.asp) recommendation
- [Google](https://google.github.io/styleguide/htmlcssguide.xml?showone=Indentation#Indentation) HTML/CSS guide

File: tpl/pluginsadmin.html [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                                <th>Name</th>
+                                <th>Description</th>
+                            </tr>
+                        </thead>
+                        <tbody>
+                            {loop="$enabled_plugins"}
+                                <tr data-line="{$key}" data-order="{$counter}">
+                                    <td class="center"><input type="checkbox" name="{$key}" checked="checked"></td>
+                                    <td class="center">
+                                        <a href="#" onclick="return orderUp(this.parentNode.parentNode.getAttribute('data-order'));">
```
Comment (virtualtam) [inline tpl/pluginsadmin.html]: XML/HTML node attributes can be spread over a couple lines to improve readability:

```html
<a href="#"
   onclick="return orderUp(this.parentNode.parentNode.getAttribute('data-order'));">▲</a>
<a href="#"
   onclick="return orderDown(this.parentNode.parentNode.getAttribute('data-order'));">▼</a>
```

File: tpl/tools.html [hunk 3]
```diff
@@ -5,11 +5,12 @@
 <div id="pageheader">
 	{include="page.header"}
 	<div id="toolsdiv">
-		{if="!$GLOBALS['config']['OPEN_SHAARLI']"}<a href="?do=changepasswd"><b>Change password</b>  <span>: Change your password.</span></a><br><br>{/if}
-		<a href="?do=configure"><b>Configure your Shaarli</b> <span>:  Change Title, timezone...</span></a><br><br>
-		<a href="?do=changetag"><b>Rename/delete tags</b> <span>:  Rename or delete a tag in all links</span></a><br><br>
-		<a href="?do=import"><b>Import</b> <span>:  Import Netscape html bookmarks (as exported from Firefox, Chrome, Opera, delicious...)</span></a> <br><br>
-		<a href="?do=export"><b>Export</b> <span>:  Export Netscape html bookmarks (which can be imported in Firefox, Chrome, Opera, delicious...)</span></a><br><br>
+		<a href="?do=configure"><b>Configure your Shaarli</b><span>:  Change Title, timezone...</span></a><br><br>
```
Comment (virtualtam) [inline tpl/tools.html]: - double space after `:`
- `<br/>`s can be put on a separate line

File: tpl/tools.html [hunk 4]
```diff
@@ -5,11 +5,12 @@
...
+		<a href="?do=pluginadmin"><b>Plugin administration</b><span>:  Enable, disable and configure plugins.</span></a><br><br>
+		{if="!$GLOBALS['config']['OPEN_SHAARLI']"}<a href="?do=changepasswd"><b>Change password</b><span>: Change your password.</span></a><br><br>{/if}
+		<a href="?do=changetag"><b>Rename/delete tags</b><span>:  Rename or delete a tag in all links</span></a><br><br>
+		<a href="?do=import"><b>Import</b><span>:  Import Netscape html bookmarks (as exported from Firefox, Chrome, Opera, delicious...)</span></a> <br><br>
```
Comment (virtualtam) [inline tpl/tools.html]: extra space before `<br/>`

File: tpl/tools.html [hunk 5]
```diff
@@ -5,11 +5,12 @@
...
+		<a href="?do=export"><b>Export</b><span>:  Export Netscape html bookmarks (which can be imported in Firefox, Chrome, Opera, delicious...)</span></a><br><br>
```
Comment (virtualtam) [inline tpl/tools.html]: double space after `:`

File: tests/ConfigTest.php [hunk 6]
```diff
@@ -174,4 +174,115 @@ public function testMergeDeprecatedConfigNoFile()
...
+            'order_plugin4' => 0,
+        );
+
+        save_plugin_config($data);
+    }
+
+    /**
+     * Test save_plugin_config without data.
+     *
+     * @throws PluginConfigOrderException
```
Comment (virtualtam) [inline tests/ConfigTest.php]: s/throws/expectedException ;-)

File: tests/ConfigTest.php [hunk 7]
```diff
@@ -174,4 +174,115 @@ public function testMergeDeprecatedConfigNoFile()
...
+            'plugin4',
+            'plugin2',
+        );
+
+        $out = save_plugin_config($data);
+        $this->assertEquals($expected, $out);
+    }
+
+    /**
+     * Test save_plugin_config with valid data.
```
Comment (virtualtam) [inline tests/ConfigTest.php]: *invalid

File: plugins/readityourself/readityourself.meta [hunk 8]
```diff
@@ -0,0 +1,2 @@
+description="Add ReadItYourself icon to save shaares URL for each links."
```
Comment (virtualtam) [inline plugins/readityourself/readityourself.meta]: Maybe rephrased the other way around? e.g. "For each link, add a ReadItYourself icon to save the shaared URL"

File: plugins/addlink_toolbar/addlink_toolbar.meta [hunk 9]
```diff
@@ -0,0 +1,2 @@
+description="Adds the addlink input on the linklist page."
+parameters=
```
Comment (virtualtam) [inline plugins/addlink_toolbar/addlink_toolbar.meta]: if an empty value is possible, could this metadata entry be optional?

File: index.php [hunk 10]
```diff
@@ -1764,6 +1764,53 @@ function renderPage()
...
+        // Split plugins into 2 arrays: ordered enabled plugins and disabled.
+        $enabledPlugins = array_filter($pluginMeta, function($v) { return $v['order'] !== false; });
+        // Load parameters.
+        $enabledPlugins = load_plugin_parameter_values($enabledPlugins, $GLOBALS['plugins']);
+        uasort($enabledPlugins, function($a, $b) {
+            return $a['order'] - $b['order'];
+        });
+        $disabledPlugins = array_filter($pluginMeta, function($v) { return $v['order'] === false; });
+
+        $PAGE->assign('enabled_plugins', $enabledPlugins);
```
Comment (virtualtam) [inline index.php]: could the camelCase syntax be kept when setting template variables?

File: index.php [hunk 11]
```diff
@@ -1764,6 +1764,53 @@ function renderPage()
...
 
+    // Plugin administration page
+    if ($targetPage == Router::$PAGE_PLUGINSADMIN) {
+        $pluginMeta = $pluginManager->getPluginsMeta();
+
+        // Split plugins into 2 arrays: ordered enabled plugins and disabled.
+        $enabledPlugins = array_filter($pluginMeta, function($v) { return $v['order'] !== false; });
+        // Load parameters.
+        $enabledPlugins = load_plugin_parameter_values($enabledPlugins, $GLOBALS['plugins']);
+        uasort($enabledPlugins, function($a, $b) {
```
Comment (virtualtam) [inline index.php]: multiline formatting suggestion:

```php
uasort(
    $enabledPlugins,
    function($a, $b) { return $a['order'] - $b['order']; }
);
```

File: index.php [hunk 12]
```diff
@@ -1764,6 +1764,53 @@ function renderPage()
...
+                foreach ($_POST as $param => $value) {
+                    $GLOBALS['plugins'][$param] = escape($value);
+                }
+            }
+            else {
+                $GLOBALS['config']['ENABLED_PLUGINS'] = save_plugin_config($_POST);
+            }
+            writeConfig($GLOBALS, isLoggedIn());
+        }
+        catch(Exception $e) {
```
Comment (virtualtam) [inline index.php]: +space after `catch`

File: inc/shaarli.css [hunk 13]
```diff
@@ -1120,4 +1120,69 @@ div.dailyNoEntry {
...
+    width: 100%;
+}
+
+#pluginsadmin table, #pluginsadmin th, #pluginsadmin td {
+    border-width: 1px 0;
+    border-style: solid;
+    border-color: #c0c0c0;
+}
+
+#pluginsadmin table tr {
```
Comment (virtualtam) [inline inc/shaarli.css]: empty rule

File: inc/plugin_admin.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+{
+    if (pos == 0) {
+        return false;
+    }
+    changeOrder(pos, -1);
+    return false;
+}
+
+/**
+ * Move a down up in the table.
```
Comment (virtualtam) [inline inc/plugin_admin.js]: s/down up/row down/

File: application/PluginManager.php [hunk 15]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+define('META_EXT', 'meta');
```
Comment (virtualtam) [inline application/PluginManager.php]: why not using a static class variable?

File: application/Config.php [hunk 16]
```diff
@@ -74,6 +74,106 @@ function writeConfig($config, $isLoggedIn)
 }
 
 /**
+ * Process plugin administration form data and save it in an array.
+ *
+ * @param $formData
```
Comment (virtualtam) [inline application/Config.php]: incomplete docstrings

**Comment by virtualtam** [comment]
First batch of comments posted ;-)

File: tpl/pluginsadmin.html (hunk 1 above)
Comment (ArthurHoaro) [inline tpl/pluginsadmin.html]: That's better!

File: tests/ConfigTest.php (hunk 6 above)
Comment (ArthurHoaro) [inline tests/ConfigTest.php]: Good catch. It wasn't actually throwing anything. Probably a copy/paste.

File: plugins/readityourself/readityourself.meta (hunk 8 above)
Comment (ArthurHoaro) [inline plugins/readityourself/readityourself.meta]: That's better. My English is potato.

File: index.php (hunk 10 above)
Comment (ArthurHoaro) [inline index.php]: Why not. I didn't use camelCase in previous PR though (most template placeholders for instance).

File: application/PluginManager.php (hunk 15 above)
Comment (ArthurHoaro) [inline application/PluginManager.php]: No idea. Changed it.

**Comment by ArthurHoaro** [comment]
Thanks for the review! I had miss a few things! ;) 
[This one](https://github.com/shaarli/Shaarli/pull/388#discussion_r49133223) is still open.

**Comment by ArthurHoaro** [comment]
Documentation updated:

  * [user guide](https://github.com/shaarli/Shaarli/wiki/Plugin-installation-%26-configuration)
  * [meta](https://github.com/shaarli/Shaarli/wiki/Plugin-System#metadata)
  * [template](https://github.com/shaarli/Shaarli/wiki/Plugin-System#plugin-administration)

**Comment by ArthurHoaro** [comment]
I just force pushed an outdated version. -.-
I'll re-fix your review and merge.
